### PR TITLE
Add templates lecture materials

### PR DESCRIPTION
**Pull request by ivankochin**

### CODE REVIEW

File: 01_templates/01_04_argument_type_deduction/01_04_00_argument_type_deduction.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+void func(T arg) { T t; }
+
+template <typename T>
+void ref_func(T& arg) { T t; }
+
+int main() {
+    int my_int{0};
+    int& my_int_ref = my_int;
+
+
```
Comment (ivankochin) [inline 01_templates/01_04_argument_type_deduction/01_04_00_argument_type_deduction.cpp:16]:
```suggestion

```

File: 01_templates/01_07_templates_and_nesting/01_07_00_templates_and_nesting.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#include <iostream>
```
Comment (pavelkumbrasev) [inline 01_templates/01_07_templates_and_nesting/01_07_00_templates_and_nesting.cpp]: File name should be : templates_and_inheritance

**Review by pavelkumbrasev — COMMENTED** [review]
Except file name, this patch looks good to me.

**Comment by ivankochin** [comment]
Add info comments

File: 01_templates/01_02_possible_template_parameters/01_02_00_possible_template_parameters.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Describe different types of template parameters
```
Comment (ivankochin) [inline 01_templates/01_02_possible_template_parameters/01_02_00_possible_template_parameters.cpp]:
```suggestion
// Describe possible template parameters
```

File: 01_templates/01_03_named_requirement/01_03_00_named_requirement.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// Check is anything works correctly
+// Talk about contracts (relations between required and provided interfaces)
+// Implement generic swap
```
Comment (ivankochin) [inline 01_templates/01_03_named_requirement/01_03_00_named_requirement.cpp]:
```suggestion
// Check if swap() works correctly
// Talk about contracts (type matching requirements)
// Implement generic swap
```

File: 01_templates/01_04_argument_type_deduction/01_04_01_deduction_absence_cases.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// Define the cases in which type deduction is not present (mention C++17 CTAD)
+
+#include <cstdlib>
+
```
Comment (ivankochin) [inline 01_templates/01_04_argument_type_deduction/01_04_01_deduction_absence_cases.cpp]:
```suggestion

```

File: 01_templates/01_05_template_specialization/01_05_00_function_template_specialization.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
+// What is template specialization
+// Differencies between full and partial specialization
+// Cons of function partial specialization
+// How to get rid of function template specialization
```
Comment (ivankochin) [inline 01_templates/01_05_template_specialization/01_05_00_function_template_specialization.cpp]:
```suggestion
// Compare function overloads with function template specializations
```

File: 01_templates/01_05_template_specialization/01_05_01_class_full_specialization.cpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
+#include <iostream>
```
Comment (ivankochin) [inline 01_templates/01_05_template_specialization/01_05_01_class_full_specialization.cpp]:
```suggestion
// Full specialization of a class template
// Full member function specialization of a class template

#include <iostream>
```

File: 01_templates/01_05_template_specialization/01_05_02_class_partial_specialization.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+template <typename U>
+struct Z<U&> {
+    Z() { std::cout << "I work with reference" << std::endl; }
+};
+
+template <>
+struct Z<char&> {
+    Z() { std::cout << "I work with char reference" << std::endl; }
+};
+
```
Comment (ivankochin) [inline 01_templates/01_05_template_specialization/01_05_02_class_partial_specialization.cpp]:
```suggestion

```

File: 01_templates/01_05_template_specialization/01_05_02_class_partial_specialization.cpp [hunk 9]
```diff
@@ -0,0 +1,40 @@
+#include <iostream>
```
Comment (ivankochin) [inline 01_templates/01_05_template_specialization/01_05_02_class_partial_specialization.cpp]:
```suggestion
// Partial template specialization
// Show specialization for reference
// Show that full specialization is preferred

#include <iostream>
```

File: 01_templates/01_06_template_inside_template/01_06_00_template_inside_template.cpp [hunk 10]
```diff
@@ -0,0 +1,17 @@
+// Is this sample compiles correctly?
+// What should we do if we want to pass any convertible type to some tempalte instance object?
```
Comment (ivankochin) [inline 01_templates/01_06_template_inside_template/01_06_00_template_inside_template.cpp]: Make it more concrete
- What is the problem
- How to add independent template

File: 01_templates/01_07_templates_and_inheritance/01_07_00_templates_and_inheritance.cpp [hunk 11]
```diff
@@ -0,0 +1,25 @@
+// How to define class that inherited from template
+// How to define template that inherited from template
+// Check knowledge of ctor calls order
```
Comment (ivankochin) [inline 01_templates/01_07_templates_and_inheritance/01_07_00_templates_and_inheritance.cpp]:
```suggestion
// How to define a class that inherited from a template
// How to define a template that inherited from a template
```

File: 01_templates/01_08_misc/01_08_01_template_aliases.cpp [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    iterator begin;
+    iterator end;
+};
+
+int main() {
+    integral_array<> my_arr1;
+    integral_array<15> my_arr2;
+
+    integral_vector::iterator it;
+}
```
Comment (ivankochin) [inline 01_templates/01_08_misc/01_08_01_template_aliases.cpp]:
```suggestion
template<typename T>
struct integral_vector {
    using iterator = T*;
    iterator begin;
    iterator end;
};

int main() {
    integral_array<> my_arr1;
    integral_array<15> my_arr2;

    integral_vector<int>::iterator it;
}

```

**Review by rarutyun — APPROVED** [review]

**Pull request merged**